initial restructuring of docs and barebones inclusion of other tools re: commercialhaskell/stack#894

kairuku opened the pull request:
-- creates new directory structure for documentation, aiming for smaller topic-based files and
   limiting repetition of content as much as possible (could potentially use table of content
   files to point to these, and potentially further nest groups of related topic files into new
   subdirectories, letting the "toc" files hide this organizational detail)

-- creates a separate doc for motivation behind stack, tools comparison, and future roadmap

-- pulls the "why" section out of the readme and the "comparison" section out of the guide,
   moving these both to the new document mentioned above

-- moves as much as possible among documentation-esque files from the root directory of the repo
   into the new docs directory, leaving only the readme in the top level directory

-- edits readme to point to the new document (URLs will need to be fixed in a subsequent commit)

-- divides docs directory into subdirectories for maintainer docs, and docs for everyone
   (inspired initially by a "release" document found elsewhere in the repo)

-- creates suggested stub directories for haddocks and man pages, but nothing has been moved here
   (one idea is that haddocks placed here with all the other documentation could be rsynced to any
   "dist" directory during any build process if needed)

-- creates suggested example stub file for wiki migration (assuming this is a desirable paradigm),
   specifically a "faq" file

kairuku commented:
(moving this comment from the "issue" to this "conversation")

I'm new to forking and pull-requests (and at the sub-novice level with Haskell too) so please bare with me :)  I didn't yet go deeply into differentiation of tools, and in this first branch have mainly just explored some ideas in creating a new directory structure for documentation (I'm happy to undo any of this too). I'll need to repair some URLs in a subsequent commit if this were to be approved for merging.

Minor notes: I wasn't sure specific markdown flavor was being targeted (I've just heretofore leaned on the syntax guide in Haroopad, but noted some difference in, e.g., the GUIDE.md file), nor whether it was desirable to break lines in the markdown files around column 80 or if this was just incidental and not necessary.

kairuku commented:
ref [#894](https://github.com/commercialhaskell/stack/issues/894)

snoyberg commented on README.md:
```diff
@@ -3,8 +3,7 @@
 
-`stack` is a cross-platform program for developing Haskell
-projects. It is aimed at Haskellers both new and experienced.
+`stack` is a cross-platform program for developing Haskell projects. It is aimed at Haskellers both new and experienced.
```
Why the collapsed line?

kairuku replied:
Sorry, I wasn't sure it was intentional, and it's just force of habit for me to remove line breaks in anything but stuff that is only likely to be read inside a small terminal window.

snoyberg commented on README.md:
```diff
@@ -16,6 +15,8 @@ It features:
 * Testing your project.
 * Benchmarking your project.
 
+`stack` replaces `stackage-cli`. For more details on the motivation for the creation of `stack`, a comparison of `stack` to other tools, and the roadmap for the future of `stack', [see this document]().
```
I'm not sure stackage-cli was ever popular enough for this message to warrant being front and center in the README, I think it will end up leading to more confusion around "what's stackage-cli?"

wolftune replied:
reference to stackage-cli looks like unneeded trivia to me that most people don't need to know at all.

kairuku replied:
Understood. I was having the opposite problem; I didn't even realize stackage-cli was deprecated until I stumbled upon its actual Hackage page, after opening #894 :)

snoyberg commented on docs/primary/topic_based_expanded_reference/GUIDE.md:
```diff
@@ -1352,30 +1352,6 @@ reminder for future uses of stack:
 * stack.yaml can override the flag settings for individual packages
 * stack.yaml specifies which packages to include
 
-## Comparison to other tools
```
I'm not certain this should be moved out of the guide. If it is:

1. @wolftune just did a significant overhaul of the text in the guide, so the updated text should be copied over.
2. We should have a link to the new file added here.

kairuku replied:
I was just trying to break things down where possible into smaller "topic" type files that could potentially be referenced from multiple places, e.g., "Installation and Upgrading" could've been another potential "topic".

snoyberg commented:
Thanks for getting started on this! I've added some inline comments, two meta comments:

1. I'm not sure if a deeply-nested directory structure is beneficial here, it may simply prevent people from finding documentation.
2. I'd like to keep a GUIDE.md in the original location, containing a link to the new location, for those already sharing the link.

kairuku commented:
@snoyberg I've scaled back the PR some

snoyberg commented:
I don't know how realistic a future merger of stack and Halcyon is, and I wouldn't want to include that without @mietek's blessing. As for comparison to other tools:

* cabal-meta inspired a lot of the multi-package functionality of stack. If you're still using cabal-install, cabal-meta is relevant. For stack work, the feature set is fully subsumed by stack.
* cabal-src is mostly irrelevant in the presence of both stack and cabal sandboxes, both of which make it easier to add additional package sources easily. The mega-sdist executable that ships with cabal-src is, however, still relevant. Its functionality may some day be folded into stack
* stackage-cli was an initial attempt to make cabal-install work more easily with curated snapshots, but due to a slight impedance mismatch between cabal.config constraints and snapshots, it did not work as well as hoped. It is deprecated in favor of stack.

Does that sound about right?

kairuku commented:
@snoyberg I think it looks great, and you know more about this than I :) I added your explanations verbatim if that is okay. Hopefully this section could be of some value to someone in the future and won't be just trivia; I cleared up my own initial confusion about stack vs stackage-cli after opening the PR.

I removed the Roadmap section because its only content current was an example link to the Google Groups discussion, but I still kind of think the general concept of a Roadmap section located adjacent to any Comparison content (or place in a separate file, linked to from any relevant section in any other file) could possibly be useful.

I could still make a /doc directory, I just wasn't sure if there was still a waiting period for comments on #900 to settle on some naming scheme, etc, but these things could also be addressed in a separate PR too. I'm not sure what anyone might have thought about (or the implications of) moving things like the man page(s), haddocks, LICENSE , CHANGELOG, etc, under /doc, it just seemed related in my mind.

snoyberg commented:
Thanks! I think a separate PR for the doc directory would be great.
